python/tskit/tables.py

```diff
@@ -2558,7 +2558,7 @@ def map_ancestors(self, *args, **kwargs):
         # A deprecated alias for link_ancestors()
         return self.link_ancestors(*args, **kwargs)
 
-    def sort(self, edge_start=0):
+    def sort(self, edge_start=0, canonical=False, unchanged_populations=False):
```

Comment:
I'm not sure this is the right way to do this. Wouldn't it be better to add a separate function ``canonicalise``, which does the sorting *and* optional filtering of unreferenced stuff?  Is there a good reason to not make a separate method? It would make the C implementation clearer I think, as well.

Reply:
Are you thinking that `canonicalize` would have a separate implementation, or would end up calling `sort`? I did it this way because there'd be a lot of code overlap.

Reply:
I'm imagining that it would look something like
```c
int TSK_WARN_UNUSED
tsk_table_collection_canonicalise(tsk_table_collection_t *self,  tsk_flags_t options)
{
    int ret = 0;
    tsk_id_t k;
    tsk_flags_t subset_options = options & TSK_UNCHANGED_POPULATIONS;
    tsk_id_t *nodes;
    tsk_table_sorter_t sorter;

    ret = tsk_table_sorter_init(&sorter, self, options);
    if (ret != 0) {
        goto out;
    }

        nodes = malloc(self->nodes.num_rows * sizeof(*nodes));
        if (nodes == NULL) {
            ret = TSK_ERR_NO_MEMORY;
            goto out;
        }
        for (k = 0; k < (tsk_id_t) self->nodes.num_rows; k++) {
            nodes[k] = k;
        }
        subset_options
            |= (TSK_SUBSET_KEEP_UNUSED_POPULATIONS | TSK_SUBSET_KEEP_UNUSED_INDIVIDUALS
                | TSK_SUBSET_KEEP_UNUSED_SITES);
        ret = tsk_table_collection_subset(
            self, nodes, self->nodes.num_rows, subset_options);
        sorter.sort_mutations = tsk_table_sorter_sort_mutations_canonical;
        tsk_safe_free(nodes);
 
    ret = tsk_table_sorter_run(&sorter, start);
    if (ret != 0) {
        goto out;
    }
out:
    tsk_table_sorter_free(&sorter);
    return ret;
}
```

I think we can add a flag to the ``tsk_table_sorter_t`` struct to decide whether we're sorting mutations in canonical order or just the standard order (i.e., I agree with you that the current duplication in tsk_table_sorter_sort_mutations_canonical isn't very nice).

And there's no changes to the current ``tsk_table_collection_sort``.

How does that sound?

Reply:
Sounds good, I had this in mind as an alternative also. I will make the change (and the other suggestions).

Reply:
> I think we can add a flag to the tsk_table_sorter_t struct to decide whether we're sorting mutations in canonical order or just the standard order (i.e., I agree with you that the current duplication in tsk_table_sorter_sort_mutations_canonical isn't very nice).

Ok, I've put back `sort` and made `canonicalise`, but I've kept `tsk_table_sorter_sort_mutations_canonical` separate from `tsk_table_sorter_sort_mutations` because the two functions use different types for `sorted_mutations`, and I don't know how to write the code in a way that removes the duplications.